### PR TITLE
Ghgc 269/fix workflows api

### PR DESCRIPTION
**Pull request by sanzog03**
**Summary:** Summary of changes

Addresses [GHGC-269: Fix workflows api for GHGC](https://github.com/US-GHG-Center/ghgc-architecture/issues/269)

## Changes

- Add support for multiple item_assets.
- Made DiscoveryItemAssets schema optional, and added a default item_assets if not provided.
- Add in missing attributes as optional, needed for stac collections and discovery_items in ghgc. Namely renderers, stac_extensions for stac collection and id_regex, id_template and assets for discovery_items.
- Fix on list dags endpoint response.
- Other minor changes and fixes.

## PR Checklist

- [ ] Update CHANGELOG
- [ ] Unit tests
- [x] Ad-hoc testing - Deploy changes and test manually
- [x] Integration tests

### CODE REVIEW

File: workflows_api/runtime/src/collection_publisher.py [hunk 1]
```diff
@@ -137,14 +140,33 @@ def create_cog_collection(self, dataset: COGDataset) -> dict:
...
+                        k: v for k, v in asset.items() if k != "regex"
+                    }
+        else:
+            collection_stac["item_assets"] = {
+                "cog_default": {
+                    "type": "image/tiff; application=geotiff; profile=cloud-optimized",
+                    "roles": ["data", "layer"],
+                    "title": "Default COG Layer",
+                    "description": "Cloud optimized default layer to display on map",
+                    "regex": "*"
```
Comment (slesaad) [inline workflows_api/runtime/src/collection_publisher.py:166]: actually, this should probably be whatever `filename_regex` is instead

Comment (sanzog03) [reply workflows_api/runtime/src/collection_publisher.py]: Updated with the mentioned change:
image

**Comment by slesaad** [comment]
some old changes are showing up as changed for some reason 🤔 

also, I wanna test it out some... so bear with me ...

**Review by slesaad — APPROVED** [review]

File: infrastructure/main.tf [hunk 2]
```diff
@@ -277,7 +303,7 @@ resource "null_resource" "update_cloudfront" {
   count = coalesce(var.cloudfront_id, false) != false ? 1 : 0
 
   provisioner "local-exec" {
-    command = "${path.module}/cf_update.sh ${var.cloudfront_id} workflows_api_origin \"${aws_apigatewayv2_api.workflows_http_api.api_endpoint}\""
+    command = "${path.module}/cf_update.sh ${var.cloudfront_id} workflows-api \"${aws_apigatewayv2_api.workflows_http_api.api_endpoint}\""
```
Comment (anayeaye) [inline infrastructure/main.tf:306]: This is a change we opted out of in https://github.com/NASA-IMPACT/veda-data-airflow/pull/167 so we should double check if we still want it

Comment (slesaad) [reply infrastructure/main.tf:306]: ah i think this branch was branched off of that branch so this sneaked in here

File: workflows_api/runtime/src/schemas.py [hunk 3]
```diff
@@ -138,19 +149,27 @@ def exists(cls, collection):
...
+                validators.collection_exists(collection_id=collection)
+            except ValueError:
+                raise ValueError(
+                    "Invalid id - id must be all lowercase, with optional '-' delimiters"
+                )
         return collection
 
 
 class S3Input(WorkflowInputBase):
-    discovery: Literal["s3"]
```
Comment (slesaad) [inline workflows_api/runtime/src/schemas.py:146]: consider leaving this?

File: infrastructure/main.tf (hunk 2 above)
Comment (slesaad) [reply infrastructure/main.tf]: removed

File: workflows_api/runtime/src/schemas.py (hunk 3 above)
Comment (slesaad) [reply workflows_api/runtime/src/schemas.py]: reverted

**Review by smohiudd — APPROVED** [review]

**Comment by slesaad** [comment]
# Testing

Case 1: No discovery_items.assets, but item_assets provided

Input:
<details>
<summary>omi-trno2-item-assets-only</summary>
```json
{
  "collection": "omi-trno2-item-assets-only",
  "data_type": "cog",
  "spatial_extent": {
    "xmin": -127,
    "ymin": 29,
    "xmax": -103,
    "ymax": 52
  },
  "temporal_extent": {
    "startdate": "1995-01-01T00:00:00Z",
    "enddate": "2095-03-31T00:00:00Z"
  },
  "description": "OMI_trno2 - 0.10 x 0.10 Annual as Cloud-Optimized GeoTIFFs (COGs)",
  "discovery_items": [
    {
      "bucket": "veda-data-store-staging",
      "datetime_range": "year",
      "discovery": "s3",
      "filename_regex": "^(.*).tif$",
      "prefix": "OMI_trno2-COG/"
    }
  ],
  "is_periodic": true,
  "license": "MIT",
  "sample_files": ["s3://veda-data-store-staging/OMI_trno2-COG/OMI_trno2_0.10x0.10_2005_Col3_V4.tif"],
  "providers": [
    {
      "name": "NASA VEDA",
      "roles": [
        "host"
      ],
      "url": "https://www.earthdata.nasa.gov/dashboard/"
    }
  ],
  "renders": {
    "dashboard": {
      "assets": [
        "cog_default"
      ],
      "colormap_name": "reds",
      "rescale": [
        [
          0,
          3000000000000000
        ]
      ],
      "title": "VEDA Dashboard Render Parameters"
    }
  },
  "item_assets": {
    "no2": {
      "type": "image/tiff; application=geotiff; profile=cloud-optimized",
      "roles": [
        "data",
        "layer"
      ],
      "title": "NO2 values",
      "description": "description",
      "other_attr": "lets see"
    }
  },
  "assets": {
    "thumbnail": {
      "description": "Photo by [Mick Truyts](https://unsplash.com/photos/x6WQeNYJC1w) (Power plant shooting steam at the sky)",
      "href": "https://thumbnails.openveda.cloud/no2--dataset-cover.jpg",
      "roles": [
        "thumbnail"
      ],
      "title": "Thumbnail",
      "type": "image/jpeg"
    }
  },
  "time_density": "year",
  "title": "DELETE ME OMI_trno2"
}
```
</details>

Output:
https://dev.openveda.cloud/api/stac/collections/omi-trno2-item-assets-only

Case 2: item_assets and discovery.assets provided
<details>
<summary>omi-trno2-custom-assets</summary>

```json
{
  "collection": "omi-trno2-custom-assets",
  "data_type": "cog",
  "spatial_extent": {
    "xmin": -127,
    "ymin": 29,
    "xmax": -103,
    "ymax": 52
  },
  "temporal_extent": {
    "startdate": "1995-01-01T00:00:00Z",
    "enddate": "2095-03-31T00:00:00Z"
  },
  "description": "OMI_trno2 - 0.10 x 0.10 Annual as Cloud-Optimized GeoTIFFs (COGs)",
  "discovery_items": [
    {
      "bucket": "veda-data-store-staging",
      "datetime_range": "year",
      "discovery": "s3",
      "filename_regex": "^(.*).tif$",
      "prefix": "OMI_trno2-COG/",
      "assets": {
        "no2": {
          "type": "image/tiff; application=geotiff; profile=cloud-optimized",
          "roles": [
            "data",
            "layer"
          ],
          "title": "NO2 values",
          "description": "description",
          "other_attr": "lets see",
          "regex": ".*"
        }
      }
    }
  ],
  "is_periodic": true,
  "license": "MIT",
  "sample_files": ["s3://veda-data-store-staging/OMI_trno2-COG/OMI_trno2_0.10x0.10_2005_Col3_V4.tif"],
  "providers": [
    {
      "name": "NASA VEDA",
      "roles": [
        "host"
      ],
      "url": "https://www.earthdata.nasa.gov/dashboard/"
    }
  ],
  "renders": {
    "dashboard": {
      "assets": [
        "cog_default"
      ],
      "colormap_name": "reds",
      "rescale": [
        [
          0,
          3000000000000000
        ]
      ],
      "title": "VEDA Dashboard Render Parameters"
    }
  },
  "assets": {
    "thumbnail": {
      "description": "Photo by [Mick Truyts](https://unsplash.com/photos/x6WQeNYJC1w) (Power plant shooting steam at the sky)",
      "href": "https://thumbnails.openveda.cloud/no2--dataset-cover.jpg",
      "roles": [
        "thumbnail"
      ],
      "title": "Thumbnail",
      "type": "image/jpeg"
    }
  },
  "item_assets": {
    "no2": {
      "type": "image/tiff; application=geotiff; profile=cloud-optimized",
      "roles": [
        "data",
        "layer"
      ],
      "title": "NO2 values",
      "description": "description",
      "other_attr": "lets see"
    }
  },
  "time_density": "year",
  "title": "DELETE ME OMI_trno2"
}
```
</details>

Output:
https://dev.openveda.cloud/api/stac/collections/omi-trno2-custom-assets

Case 3: No assets anywhere
<details>
<summary>omi-trno2-no-assets</summary>

```json
{
  "collection": "omi-trno2-no-assets",
  "data_type": "cog",
  "spatial_extent": {
    "xmin": -127,
    "ymin": 29,
    "xmax": -103,
    "ymax": 52
  },
  "temporal_extent": {
    "startdate": "1995-01-01T00:00:00Z",
    "enddate": "2095-03-31T00:00:00Z"
  },
  "description": "OMI_trno2 - 0.10 x 0.10 Annual as Cloud-Optimized GeoTIFFs (COGs)",
  "discovery_items": [
    {
      "bucket": "veda-data-store-staging",
      "datetime_range": "year",
      "discovery": "s3",
      "filename_regex": "^(.*).tif$",
      "prefix": "OMI_trno2-COG/"
    }
  ],
  "is_periodic": true,
  "license": "MIT",
  "sample_files": ["s3://veda-data-store-staging/OMI_trno2-COG/OMI_trno2_0.10x0.10_2005_Col3_V4.tif"],
  "providers": [
    {
      "name": "NASA VEDA",
      "roles": [
        "host"
      ],
      "url": "https://www.earthdata.nasa.gov/dashboard/"
    }
  ],
  "renders": {
    "dashboard": {
      "assets": [
        "cog_default"
      ],
      "colormap_name": "reds",
      "rescale": [
        [
          0,
          3000000000000000
        ]
      ],
      "title": "VEDA Dashboard Render Parameters"
    }
  },
  "assets": {
    "thumbnail": {
      "description": "Photo by [Mick Truyts](https://unsplash.com/photos/x6WQeNYJC1w) (Power plant shooting steam at the sky)",
      "href": "https://thumbnails.openveda.cloud/no2--dataset-cover.jpg",
      "roles": [
        "thumbnail"
      ],
      "title": "Thumbnail",
      "type": "image/jpeg"
    }
  },
  "time_density": "year",
  "title": "DELETE ME OMI_trno2"
}
```
</details>

Output:
https://dev.openveda.cloud/api/stac/collections/omi-trno2-no-assets

Case 4: multi assets
<details>
<summary>climdex-tmaxxf-access-cm2-ssp126-multi-asset</summary>

```json
{ 
  "collection": "climdex-tmaxxf-access-cm2-ssp126-multi-asset",
  "data_type": "cog",
  "spatial_extent": {
    "xmin": -180,
    "ymin": -90,
    "xmax": 180,
    "ymax": 90
  },
  "temporal_extent": {
    "startdate": "2015-01-01T00:00:00Z",
    "enddate": "2101-12-31T23:59:59Z"
  },
  "description": "CLIMDEX ACCESS CM2 SSP125 - variable tmaxXF",
  "is_periodic": true,
  "license": "MIT",
  "item_assets": {
    "cog_default": {
        "type": "image/tiff; application=geotiff; profile=cloud-optimized",
        "roles": [
            "data",
            "layer"
        ],
        "title": "Default COG Layer",
        "description": "Cloud optimized default layer to display on map"
    },
    "tmax_above_86": {
        "type": "image/tiff; application=geotiff; profile=cloud-optimized",
        "roles": [
            "data",
            "layer"
        ],
        "title": "Tmax Above 86",
        "description": "Tmax Above 86"
    },
    "tmax_above_90": {
        "type": "image/tiff; application=geotiff; profile=cloud-optimized",
        "roles": [
            "data",
            "layer"
        ],
        "title": "Tmax Above 90",
        "description": "Tmax Above 90"
    },
    "tmax_above_100": {
        "type": "image/tiff; application=geotiff; profile=cloud-optimized",
        "roles": [
            "data",
            "layer"
        ],
        "title": "Tmax Above 100",
        "description": "Tmax Above 100"
    },
    "tmax_above_110": {
        "type": "image/tiff; application=geotiff; profile=cloud-optimized",
        "roles": [
            "data",
            "layer"
        ],
        "title": "Tmax Above 110",
        "description": "Tmax Above 110"
    },
    "tmax_above_115": {
        "type": "image/tiff; application=geotiff; profile=cloud-optimized",
        "roles": [
            "data",
            "layer"
        ],
        "title": "Tmax Above 115",
        "description": "Tmax Above 115"
    }
  },
  "sample_files": ["s3://veda-data-store-staging/climdex-tmaxxf-access-cm2-ssp126/tmaxXF-ACCESS-CM2-ssp126_2099_tmax_above_86.tif"],
  "providers": [
        {
            "name": "NASA VEDA",
            "url": "https://www.earthdata.nasa.gov/dashboard/",
            "roles": [
                "host"
            ]
        }
    ],
  "renders": {
    "dashboard": {
      "assets": [
        "cog_default"
      ],
      "colormap_name": "reds",
      "rescale": [
        [
          0,
          3000000000000000
        ]
      ],
      "title": "VEDA Dashboard Render Parameters"
    }
  },
  "assets": {
    "thumbnail": {
      "title": "Thumbnail",
      "description": "Photo by NASA (CMIP6 Climdex TmaxXF Screenshot)",
      "href": "https://thumbnails.openveda.cloud/cmip6-climdex-tmaxxf-access-cm2.png",
      "type": "image/png",
      "roles": ["thumbnail"]
    }
  },
  "time_density": "year",
  "title": "DELETE ME CLIMDEX",
  "discovery_items": [
    {
      "collection": "climdex-tmaxxf-access-cm2-ssp126-deleteme",
      "bucket": "veda-data-store-staging",
      "prefix": "climdex-tmaxxf-access-cm2-ssp126/",
      "filename_regex": ".*-ssp126_209(.*)_tmax.*.tif$",
      "id_regex": ".*-ssp126_(.*)_tmax.*.tif$",
      "id_template": "climdex-tmaxxf-access-cm2-ssp126-{}",
      "datetime_range": "year",
      "assets": {
          "tmax_above_86": {
            "title": "Tmax Above 86",
            "description": "Tmax Above 86",
            "regex": ".*-ssp126_(.*)_tmax_above_86.tif"
          },
          "tmax_above_90": {
            "title": "Tmax Above 90",
            "description": "Tmax Above 90",
            "regex": ".*-ssp126_(.*)_tmax_above_90.tif"
          },
          "tmax_above_100": {
            "title": "Tmax Above 100",
            "description": "Tmax Above 100",
            "regex": ".*-ssp126_(.*)_tmax_above_100.tif"
          },
          "tmax_above_110": {
            "title": "Tmax Above 110",
            "description": "Tmax Above 110",
            "regex": ".*-ssp126_(.*)_tmax_above_110.tif"
          },
          "tmax_above_115": {
            "title": "Tmax Above 115",
            "description": "Tmax Above 115",
            "regex": ".*-ssp126_(.*)_tmax_above_115.tif"
          }
        },
      "discovery": "s3",
      "upload": false
    }
  ]
}
```
</details>
Output:
https://dev.openveda.cloud/api/stac/collections/climdex-tmaxxf-access-cm2-ssp126-multi-asset

File: workflows_api/runtime/src/main.py [hunk 4]
```diff
@@ -105,16 +105,9 @@ async def publish_dataset(
...
-                    "cog_default": {
-                        "type": "image/tiff; application=geotiff; profile=cloud-optimized",
-                        "roles": ["data", "layer"],
-                        "title": "Default COG Layer",
-                        "description": "Cloud optimized default layer to display on map",
-                        "regex": discovery_dict.get("filename_regex") or ".*"
-                    }
-                }
-            response = await start_discovery_workflow_execution(discovery)
+                if (dataset.item_assets):
```
Comment (slesaad) [inline workflows_api/runtime/src/main.py:108]: Use item_assets is no assets provided, and if that's also not provided send nothing to airflow - which handles adding the cog_default asset there

**Comment by anayeaye** [comment]
I think we need an exclude_unset=True somewhere in the dataset path to publishing a collection record to pgstac, maybe as the [last step before posting to the ingest API](https://github.com/NASA-IMPACT/veda-data-airflow/blob/GHGC-269/fix-workflows_api/workflows_api/runtime/src/collection_publisher.py#L198)? I see some published `keywords: null, stac_extensions: null, ...`

**Comment by anayeaye** [comment]
> I think we need an exclude_unset=True somewhere in the dataset path to publishing a collection record to pgstac, maybe as the [last step before posting to the ingest API](https://github.com/NASA-IMPACT/veda-data-airflow/blob/GHGC-269/fix-workflows_api/workflows_api/runtime/src/collection_publisher.py#L198)? I see some published `keywords: null, stac_extensions: null, ...`

I think it is just a switch to the stac_pydantic to_dict method which defaults exclude_unset=True
`collection.json(by_alias=True)`-->`collection.to_dict(by_alias=True)`

File: workflows_api/runtime/src/collection_publisher.py [hunk 5]
```diff
@@ -183,20 +181,4 @@ def generate_stac(
...
-        headers = {
-            "Authorization": f"Bearer {token}",
-            "Content-Type": "application/json",
-        }
-        response = requests.post(url, data=collection.json(by_alias=True), headers=headers)
-        if response.status_code == 200:
-            print("Success:", response.json())
-        else:
-            print("Error:", response.status_code, response.text)
+
```
Comment (slesaad) [inline workflows_api/runtime/src/collection_publisher.py:184]: There were apparently two `ingest` methods defined - that tripped me up for a while 😓

File: workflows_api/runtime/src/schemas.py [hunk 6]
```diff
@@ -57,7 +57,6 @@ class DashboardCollection(Collection):
     assets: Optional[Dict]
     extent: SpatioTemporalExtent
     renders: Optional[Dict]
-    stac_extensions: Optional[List[str]]
```
Comment (slesaad) [inline workflows_api/runtime/src/schemas.py]: `stac_pydantic.Collection` already defines this

File: workflows_api/runtime/src/collection_publisher.py [hunk 7]
```diff
@@ -27,8 +25,8 @@ def ingest(self, collection: DashboardCollection, token: str, ingest_api: str):
             "Authorization": f"Bearer {token}",
             "Content-Type": "application/json",
         }
-        response = requests.post(url, data=collection.json(by_alias=True), headers=headers)
-        if response.status_code == 200:
+        response = requests.post(url, data=collection.json(by_alias=True, exclude_unset=True, exclude_none=True), headers=headers)
```
Comment (slesaad) [inline workflows_api/runtime/src/collection_publisher.py:28]: `exclude_unset` didn't work for `stac_extensions` - had to add `exclude_none` too

File: workflows_api/runtime/src/collection_publisher.py (hunk 5 above)
Comment (anayeaye) [reply workflows_api/runtime/src/collection_publisher.py:184]: oh no! 🙃

**Review by anayeaye — COMMENTED** [review]
One change to id_template default value + I am still looking at whether these fields should be required or if the preference was to default them when not provided (from datasets/publish). Either way I think we can open a new issue to address this error.

```
  {'loc': ('body', 'COGDataset', 'spatial_extent'), 'msg': 'field required', 'type': 'value_error.missing'}
  {'loc': ('body', 'COGDataset', 'temporal_extent'), 'msg': 'field required', 'type': 'value_error.missing'}, 
  {'loc': ('body', 'COGDataset', 'sample_files'), 'msg': 'field required', 'type': 'value_error.missing'}
```

File: workflows_api/runtime/src/schemas.py [hunk 8]
```diff
@@ -151,6 +167,9 @@ class S3Input(WorkflowInputBase):
     start_datetime: Optional[datetime]
     end_datetime: Optional[datetime]
     single_datetime: Optional[datetime]
+    id_regex: Optional[str]
+    id_template: Optional[str]
```
[the remark this answers is not in the document]
Comment (slesaad) [reply workflows_api/runtime/src/schemas.py:171]: But I fixed it using `exclude_unset=True` when passing over the discovery config 🤔 
and when it's not set it defaults = "{}" in [this line](https://github.com/NASA-IMPACT/veda-data-airflow/blob/dev/dags/veda_data_pipeline/utils/s3_discovery.py#L200)

Did you still get this error?

[the remark this answers is not in the document]
Comment (anayeaye) [reply workflows_api/runtime/src/schemas.py:171]: For this discovery/ request
```
{
  "bucket": "veda-data-store-staging",
  "collection": "OMI_trno2-COG-deleteme-ak",
  "datetime_range": "year",
  "discovery": "s3",
  "filename_regex": "^(.*).tif$",
  "prefix": "OMI_trno2-COG/",
  "zarr_store": null
}
```
This error was resolved by defaulting to "{}"
```
AIRFLOW_CTX_DAG_OWNER=airflow
AIRFLOW_CTX_DAG_ID=veda_discover
AIRFLOW_CTX_TASK_ID=subdag_discover.discover_from_s3
AIRFLOW_CTX_EXECUTION_DATE=2024-07-19T15:53:42+00:00
AIRFLOW_CTX_TRY_NUMBER=1
AIRFLOW_CTX_DAG_RUN_ID=d222b047-d453-4980-acad-f40d473320c6
[2024-07-19, 15:53:50 UTC] {{logging_mixin.py:137}} INFO - Getting S3 response iterator for bucket: veda-data-store-staging, prefix: OMI_trno2-COG/
[2024-07-19, 15:53:50 UTC] {{taskinstance.py:1768}} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/usr/local/airflow/.local/lib/python3.10/site-packages/airflow/operators/python.py", line 247, in execute
    condition = super().execute(context)
  File "/usr/local/airflow/.local/lib/python3.10/site-packages/airflow/operators/python.py", line 175, in execute
    return_value = self.execute_callable()
  File "/usr/local/airflow/.local/lib/python3.10/site-packages/airflow/operators/python.py", line 192, in execute_callable
    return self.python_callable(*self.op_args, **self.op_kwargs)
  File "/usr/local/airflow/dags/veda_data_pipeline/groups/discover_group.py", line 36, in discover_from_s3_task
    return s3_discovery_handler(
  File "/usr/local/airflow/dags/veda_data_pipeline/utils/s3_discovery.py", line 251, in s3_discovery_handler
    item["item_id"] = id_template.format(item["item_id"])
AttributeError: 'NoneType' object has no attribute 'format'
[2024-07-19, 15:53:50 UTC] {{taskinstance.py:1318}} INFO - Marking task as FAILED. dag_id=veda_discover, task_id=subdag_discover.discover_from_s3, execution_date=20240719T155342, start_date=20240719T155349, end_date=20240719T155350
[2024-07-19, 15:53:50 UTC] {{standard_task_runner.py:100}} ERROR - Failed to execute job 2754 for task subdag_discover.discover_from_s3 ('NoneType' object has no attribute 'format'; 24409)
[2024-07-19, 15:53:50 UTC] {{local_task_job.py:208}} INFO - Task exited with return code 1
```

[the remark this answers is not in the document]
Comment (anayeaye) [reply workflows_api/runtime/src/schemas.py:171]: I'm OK deferring the change, technically the discovery endpoint path will work when an id_template is supplied in the request. I'm going to open an issue to make some dataset parameters optional and I can also create another for this id_template condition.

[the remark this answers is not in the document]
Comment (anayeaye) [reply workflows_api/runtime/src/schemas.py:171]: @slesaad I don't see why `id_template = event.get("id_template", "{}")` in the s3_discovery_handler that you pointed to isn't cutting it. It looks like it should have corrected the request without id_template.

**Pull request merged**